platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/internal/cc/impl/ConfigurationCacheStringDeduplicationIntegrationTest.groovy

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl
+
+import org.gradle.api.services.BuildService
+import org.gradle.api.services.BuildServiceParameters
+
+class ConfigurationCacheStringDeduplicationIntegrationTest extends AbstractConfigurationCacheIntegrationTest {
+
+    def "strings are deduplicated across projects"() {
+        given:
+        createDirs 'foo', 'bar'
+        settingsFile '''
+            include 'foo', 'bar'
+        '''
+        buildFile """
+            abstract class StringDedupCheckerService implements ${BuildService.name}<${BuildServiceParameters.name}.None>{
+
+                private String string = null
+
+                synchronized def check(String s) {
+                    if (string === null) {
+                        string = s
+                    } else {
+                        assert string === s
+                        println 'Strings have been deduplicated'
+                    }
+                }
+            }
+
+            abstract class StringTask extends DefaultTask {
+                @Input abstract Property<String> getString()
+                @ServiceReference('stringChecker') abstract Property<StringDedupCheckerService> getService()
+                @TaskAction def check() {
+                    service.get().check(string.get())
+                }
+            }
+
+            def service = gradle.sharedServices.registerIfAbsent('stringChecker', StringDedupCheckerService) {}
+            allprojects {
```

Comment:
❓  Is this test IP-incompatible? It can be a settings script that uses the new lifecycle API.
❓ With parallel project saving, will string deduplication work across project? Or is the goal to have it across projects?

Reply:
> Is this test IP-incompatible? It can be a settings script that uses the new lifecycle API.
Right, good point.

> With parallel project saving, will string deduplication work across project? Or is the goal to have it across projects?

Yes, that would be the idea. We discussed having all the strings handled by a separate _String Pool_ worker, let's see if we can make that fly.